diff --git a/sdk/android/BUILD.gn b/sdk/android/BUILD.gn
index 8a3e20c4e9..b4c7479293 100644
--- a/sdk/android/BUILD.gn
+++ b/sdk/android/BUILD.gn
@@ -124,11 +124,21 @@ if (is_android) {
     ]
   }
 
+  config("force_page_size_16k") {
+    ldflags = [
+      "-Wl,-z,max-page-size=16384",
+      "-Wl,-z,common-page-size=16384",
+    ]
+  }
+
   rtc_shared_library("libjingle_peerconnection_so") {
     sources = [ "src/jni/jni_onload.cc" ]
 
     suppressed_configs += [ "//build/config/android:hide_all_but_jni_onload" ]
-    configs += [ "//build/config/android:hide_all_but_jni" ]
+    configs += [
+      "//build/config/android:hide_all_but_jni",
+      ":force_page_size_16k",
+    ]
     ldflags = [
       "-lEGL",
       "-Wl,--build-id",
